Sources/ArgumentParser/DialogBuilder/Ask.swift

```diff
@@ -0,0 +1,168 @@
+//===----------------------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// Ask single string from the user after displaying a prompt.
+///
+/// ```swift
+/// let name = ask("? Please enter your name: ") { name in
+///   if name.allSatisfy({ $0 == " " }) {
+///     return .failure(.init("Error: Name cannot be empty.\n"))
+///   } else {
+///     return .success(())
+///   }
+/// }
+/// ```
+/// `validate` method can validate user input after parsing:
+///
+/// ```
+/// ? Please enter your name:
+/// Error: Name cannot be empty.
+///
+/// ? Please enter your name: keith
+/// ```
+///
+/// - Parameters:
+///   - prompt: The message to display.
+///   - getInput: Get input from the user's typing or other ways.
+///   - validate: Custom validation for user input.
+/// - Returns: The string casted to the type requested.
+internal func ask(
+  _ prompt: String,
+  getInput: () -> String? = { readLine() },
+  validate: ((String) -> Result<Void, ValidationError>)? = nil
+) -> String {
```

Comment:
I think this method could pass its own arguments, plus `type: String.self` into the next `ask(...)` function — does that let you eliminate this duplication?

Reply:
done